### DISCUSSION TITLE
Email Verification  for Imported Accounts

### DISCUSSION THREAD

**Discussion by i-h8-github** [jeremyevans/rodauth#457] [General]
Hello,

I am migrating an old WordPress site to Roda/Sequel/Rodauth, and I have to migrate a bunch of user accounts. In the process, I'd like to enforce email verification and password resets for these old accounts to weed out the many, many bots.

I was sort of hoping that merely importing the usernames/email addresses to the `accounts` table would allow the framework to gracefully handle the rest by virtue of no other information being present in the database, but unfortunately that is not so. The first roadblock is that email verification cannot be enforced because the system is unable to "resend" a verification email that it never sent in the first place.

Is there any advice on the best way to accomplish this?

Thank you for your time.

**Comment by jeremyevans** [comment]
Assuming you imported the accounts with unverified status, try this patch:

```diff
diff --git a/lib/rodauth/features/verify_account.rb b/lib/rodauth/features/verify_account.rb
index a27b857..f5e3b5c 100644
--- a/lib/rodauth/features/verify_account.rb
+++ b/lib/rodauth/features/verify_account.rb
@@ -60,6 +60,7 @@ module Rodauth
     )
 
     internal_request_method(:verify_account_resend)
+    internal_request_method(:setup_account_verification)
     internal_request_method
 
     route(:verify_account_resend) do |r|
```

And this code:

```ruby
require 'rodauth'
rodauth = Rodauth.lib do
  enable :verify_account
  # other rodauth configuration here
end
DB[:accounts].select_order_map(:id).each do |id|
  rodauth.setup_account_verification(account_id: id)
end 
```

If that works, I can commit the patch.  I haven't tried it, so it may not work.

**Reply by i-h8-github** [reply]
Thank you for your reply. I tried adding that line to the appropriate file in my rodauth installation, but it yields the following error:

```
…/lib/rodauth/features/internal_request.rb:110:in `block (2 levels) in handle_internal_request': undefined method `_handle_setup_account_verification' for #<#<Class:0x00007344eeeb3f10>::InternalRequest:0x00007344ef28f080 @scope=#<#<Class:0x00007344eeeb7c28>:0x00007344ef28f940 @_request=#<#<Class:0x00007344eeeb7c28>::RodaRequest POST />, @_response=#<#<Class:0x00007344eeeb7c28>::RodaResponse nil {} []>>, @session={:account_id=>1, :authenticated_by=>["internal1", "internal2"]}, @params={}, @internal_request_block=nil> (NoMethodError)

          send(meth, request)
          ^^^^
Did you mean?  _handle_create_account
```

**Reply by jeremyevans** [reply]
Looks like more is needed.  This diff includes a spec, so it hopefully will work (or at least be less broken):

```diff
diff --git a/lib/rodauth/features/internal_request.rb b/lib/rodauth/features/internal_request.rb
index c7224a0..3985e49 100644
--- a/lib/rodauth/features/internal_request.rb
+++ b/lib/rodauth/features/internal_request.rb
@@ -247,6 +247,11 @@ module Rodauth
       _handle_remember(request)
     end
 
+    def _handle_setup_account_verification(_)
+      @account = account_ds(session_value).first
+      setup_account_verification
+    end
+
     def _handle_account_id_for_remember_key(request)
       load_memory
       raise InternalRequestError, "invalid remember key"
diff --git a/lib/rodauth/features/verify_account.rb b/lib/rodauth/features/verify_account.rb
index a27b857..f5e3b5c 100644
--- a/lib/rodauth/features/verify_account.rb
+++ b/lib/rodauth/features/verify_account.rb
@@ -60,6 +60,7 @@ module Rodauth
     )
 
     internal_request_method(:verify_account_resend)
+    internal_request_method(:setup_account_verification)
     internal_request_method
 
     route(:verify_account_resend) do |r|
diff --git a/spec/verify_account_spec.rb b/spec/verify_account_spec.rb
index 812b903..dabe948 100644
--- a/spec/verify_account_spec.rb
+++ b/spec/verify_account_spec.rb
@@ -493,5 +493,15 @@ describe 'Rodauth verify_account feature' do
 
     login(:login=>'foo@example5.com')
     page.body.must_include 'Logged In'
+
+    ds = DB[:accounts].where(:email=>'foo@example5.com')
+    account_id = ds.get(:id)
+    ds.update(:status_id=>1)
+    DB[:account_verification_keys].delete
+    app.rodauth.setup_account_verification(:account_id=>account_id)
+
+    link = email_link(/(\/verify-account\?key=.+)$/, 'foo@example5.com')
+    key = link.split('=').last
+    app.rodauth.verify_account(:verify_account_key=>key, :password=>'0123456789').must_be_nil
   end
 end

```

**Comment by i-h8-github** [comment]
Now we are at:

```
Error: Rodauth::InternalRequestError: must set domain in configuration, as it cannot be determined from internal request
```

I'm not sure exactly what I'm intended to set for what I assume is the `domain` method of the plugin config.

**Reply by janko** [reply]
You set it to your application domain. So, if you're hosting your app at `https://my-app.com`, you'd set:

```rb
Rodauth.lib do
  enable :verify_account
  domain "my-app.com"
end
```

This is used for things like generating email links.

**Reply by i-h8-github** [reply]
That's about what I figured, but what do I set for local testing? Setting it to "localhost" yields `Connection refused`.

**Reply by janko** [reply]
I think this is because by default Rodauth's `internal_request` feature will assume `https://`, which won't work unless you have SSL certificates set up for localhost. Try overriding `base_url` as well, which accepts a full URL, where you can specify the scheme:

```rb
Rodauth.lib do
  # ...
  domain "localhost:<port>"
  base_url "http://localhost:<port>"
end
```

**Reply by i-h8-github** [reply]
I can't seem to figure out the correct incantation. Puma is serving this app on 9292, but I keep getting:

```
…/gems/net-smtp-0.4.0/lib/net/smtp.rb:631:in `initialize': Connection refused - connect(2) for "localhost" port 25 (Errno::ECONNREFUSED)
```
No combination of `localhost` or `127.0.0.1` or any port number I've tried seems to surmount this error.

Come to think of it, in production, I'd be serving this via puma behind an apache reverse proxy, so wouldn't I have the same issue there?

**Reply by jeremyevans** [reply]
That isn't a web request error, it's trying to send email to localhost, and you aren't running an SMTP server.  You'll need to use `Mail.defaults` to configure the mail gem.  Go over the mail README to see how you could configure it for your environment: https://github.com/mikel/mail

**Reply by i-h8-github** [reply]
Yeah, sorry, I forgot to mention that I have the following in `config.ru`:

```ruby
Mail.defaults do
  delivery_method :smtp, :address => "127.0.0.1", :port => 1025
end
```

So I can monitor mail with mailcatcher.

**Reply by jeremyevans** [reply]
Can you post a full backtrace for the error?

**Reply by jeremyevans** [reply]
Also, you mentioned this was in `config.ru`.  Are you sure that `config.ru` is being loaded?

**Reply by i-h8-github** [reply]
> Can you post a full backtrace for the error?

https://paste.debian.net/hidden/65d26cf1/

> Also, you mentioned this was in `config.ru`. Are you sure that `config.ru` is being loaded?

This setup had been working previously, and the only changes I've made to the working project are those discussed in this thread.

**Reply by jeremyevans** [reply]
Any chance your `Mail.defaults` setting is after line 11 in your `config.ru`?

**Reply by i-h8-github** [reply]
> Any chance your `Mail.defaults` setting is after line 11 in your `config.ru`?

Indeed, but I can't move it from the bottom or else Mail is uninitialized.

config.ru:

```ruby
dev = ENV['RACK_ENV'] == 'development'

if dev
  require 'logger'
  logger = Logger.new($stdout)
end

require 'rack/unreloader'
Unreloader = Rack::Unreloader.new(subclasses: %w'Roda Sequel::Model', logger: logger, reload: dev, autoload: dev){MapleApp}
require_relative 'models'
Unreloader.require('app.rb'){'MapleApp'}
run(dev ? Unreloader : Library.freeze.app)

unless dev
  require 'tilt/sass' unless File.exist?(File.expand_path('../compiled_assets.json', __FILE__))
  Tilt.finalize!
end

freeze_core = false
#freeze_core = !dev # Uncomment to enable refrigerator
if freeze_core
  begin
    require 'refrigerator'
  rescue LoadError
  else

    # When enabling refrigerator, you may need to load additional
    # libraries before freezing the core to work correctly.  You'll
    # want to uncomment the appropriate lines below if you run into
    # problems after enabling refrigerator.

    # rackup -s webrick
    #require 'forwardable'
    #require 'webrick'

    # rackup -s Puma
    #require 'yaml' 
    #Gem.ruby

    # Puma (needed for state file)
    #require 'yaml'

    # Unicorn (no changes needed)

    Refrigerator.freeze_core
  end
end

case ENV['RACK_ENV']
when 'development'
  Mail.defaults do
    delivery_method :smtp, :address => "127.0.0.1", :port => 1025 # or whatever port maildev SMTP listens on
  end
when 'test'
  Mail.defaults do
    delivery_method :test # this allows you to check Mail::TestMailer.deliveries for received email in tests
  end
end
```

**Reply by jeremyevans** [reply]
Your error is that `Mail.defaults` is not being called, because the error occurs at line 11 of `config.ru`.  You should `require 'mail'` early in `config.ru`, and move the `Mail.defaults` code above the `Unreloader.require('app.rb'){'MapleApp'}` line.

**Reply by i-h8-github** [reply]
Thank you. Though I do wonder why I didn't have to explicitly `require 'mail'` before.

In any case, everything does appear to be working now.

Though it has since occurred to me that importing the accounts as "Verified" and forcing a password reset essentially accomplishes the same thing, and rodauth does gracefully handle non-existent passwords.
